### PR TITLE
Use arrow kernels for bitwise operations

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/6066

# Rationale for this change

While reviewing https://github.com/apache/arrow-datafusion/pull/6093 from @RTEnzyme I noticed there were kernels in arrow-rs for some of the bitwise operations: https://docs.rs/arrow/latest/arrow/compute/kernels/bitwise/index.html

# What changes are included in this PR?

Use the kernels from arrow-rs rather than redefining them in datafusion

# Are these changes tested?

Covered by existing tests

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -5117,31 +5117,31 @@ mod tests {
         let left = Arc::new(Int32Array::from(vec![Some(12), None, Some(11)])) as ArrayRef;
         let right =
             Arc::new(Int32Array::from(vec![Some(1), Some(3), Some(7)])) as ArrayRef;
-        let mut result = bitwise_and(left.clone(), right.clone())?;
+        let mut result = bitwise_and_dyn(left.clone(), right.clone())?;
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:5120]: renamed these kernels to match the naming scheme in arrow-rs (where `_dyn` is used to refer to kernels that take `ArrayRef`

File: datafusion/physical-expr/src/expressions/binary/kernels.rs [hunk 2]
```diff
@@ -56,41 +60,64 @@ macro_rules! binary_bitwise_array_scalar {
...
-    }
+    };
 }
 
-pub(crate) fn bitwise_shift_right(left: ArrayRef, right: ArrayRef) -> Result<ArrayRef> {
+create_dyn_kernel!(bitwise_or_dyn, bitwise_or);
+create_dyn_kernel!(bitwise_xor_dyn, bitwise_xor);
+create_dyn_kernel!(bitwise_and_dyn, bitwise_and);
+
+pub(crate) fn bitwise_shift_right_dyn(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary/kernels.rs:117]: Sadly there are no shift_right or shift_left in arrows

File: datafusion/physical-expr/src/expressions/binary/kernels.rs [hunk 3]
```diff
@@ -238,333 +268,52 @@ pub(crate) fn bitwise_shift_left(left: ArrayRef, right: ArrayRef) -> Result<Arra
...
-        }
-        DataType::UInt16 => {
-            binary_bitwise_array_scalar!(
-                array,
-                scalar,
-                |a: u16| move |b: u16| a | b,
-                UInt16Array,
-                u16
-            )
+/// Creates a $FUNC(left: ArrayRef, right: ArrayRef) that
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/binary/kernels.rs:287]:
```suggestion
/// Creates a $FUNC(left: ArrayRef, right: ScalarValue) that
```

**Review by viirya — APPROVED** [review]

File: datafusion/physical-expr/src/expressions/binary/kernels.rs (hunk 2 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary/kernels.rs:118]: I also found the upstream tracking ticket in arrow -- https://github.com/apache/arrow-rs/issues/2741 and added links in the code here

**Pull request merged**